File: ruby/lib/oppex_sdk/internal/async_dispatcher.rb
# frozen_string_literal: true

require_relative "rate_limited_drop_logger"

module Oppex
  module Internal
    # Delivers asynchronous incidents on a fixed number of worker threads,
    # queueing the rest up to +capacity+ and dropping the oldest entry once full.
    #
    # Dropping the oldest keeps the newest incident, which is the one most likely
    # to still matter, and submission never blocks the application.
    #
    # Not public API.
    class AsyncDispatcher
      QUEUE_CAPACITY = 5000
      WORKER_COUNT = 2
      CLOSE_DRAIN_TIMEOUT_SECONDS = 10.0

      def initialize(logger, worker_count: WORKER_COUNT, capacity: QUEUE_CAPACITY)
        @logger = logger
        @capacity = capacity
        # One mutex guards the queue and every lifecycle flag together. Splitting
        # them would reintroduce the missed-wakeup race: a worker reads
        # "draining", close clears it and broadcasts, and only then does the
        # worker start waiting, on a broadcast that already happened.
        @mutex = Mutex.new
        @changed = ConditionVariable.new
        @tasks = []
        @accepting = true
        @draining = true
        @active = 0
        @drop_logger = RateLimitedDropLogger.new
        @workers = Array.new(worker_count) { Thread.new { work } }
      end

      # Queues a task, evicting the oldest queued task when the queue is full.
      # Returns whether the task was accepted; only a closed dispatcher refuses.
      #
      # rubocop:disable Naming/PredicateMethod -- this is a command that reports
      # whether it was accepted, not a question; `submit?` would read as one.
      def submit(&task)
        dropped = @mutex.synchronize do
          next :refused unless @accepting

          evicted = if @tasks.length >= @capacity
                      @tasks.shift
                      @drop_logger.record_drop
                    end
          @tasks.push(task)
          @changed.broadcast
          evicted
        end
        return false if dropped == :refused

        @logger.warn("oppex: dropped #{dropped} incidents in the last minute") if dropped
        true
      end
      # rubocop:enable Naming/PredicateMethod

      # Stops admitting work and drains what is already queued and in flight for
      # up to +timeout_seconds+, then abandons the rest. Idempotent, and safe to
      # call from any thread.
      def close(timeout_seconds = CLOSE_DRAIN_TIMEOUT_SECONDS)
        was_accepting = @mutex.synchronize do
          accepting = @accepting
          @accepting = false
          accepting
        end
        return unless was_accepting

        abandoned = drain_until(monotonic_now + timeout_seconds)

        # The workers are deliberately not joined without a bound. A Ruby thread
        # cannot be interrupted safely, so waiting on one that is blocked in an
        # HTTP attempt would push close past its own timeout. Java makes the same
        # trade: it waits for the drain, calls shutdownNow, and returns without
        # waiting for whatever that failed to interrupt.
        @workers.each { |worker| worker.join(0) }

        # Reported once and directly, rather than through the rate-limited
        # overflow counter: a shutdown loss and an overload loss are different
        # events, and a rate-limited counter's last batch can go unreported.
        @logger.warn("oppex: force-dropped #{abandoned} pending incidents during close") if abandoned.positive?
      end

      private

      def work
        loop do
          task = @mutex.synchronize do
            while @tasks.empty?
              return unless @draining

              @changed.wait(@mutex)
            end
            @active += 1
            @tasks.shift
          end

          begin
            task.call
          ensure
            @mutex.synchronize do
              @active -= 1
              @changed.broadcast
            end
          end
        end
      end

      # Waits until nothing is queued or in flight, or the deadline passes. Then
      # stops the workers and returns how many tasks were abandoned.
      def drain_until(deadline)
        @mutex.synchronize do
          until @tasks.empty? && @active.zero?
            remaining = deadline - monotonic_now
            break if remaining <= 0

            @changed.wait(@mutex, remaining)
          end

          @draining = false
          @changed.broadcast
          abandoned = @tasks.length
          @tasks.clear
          abandoned
        end
      end

      def monotonic_now
        Process.clock_gettime(Process::CLOCK_MONOTONIC)
      end
    end
  end
end

File: cpp/src/async_dispatcher.cpp
#include "async_dispatcher.hpp"

#include <cstdint>
#include <optional>
#include <string>
#include <utility>

namespace oppex {

AsyncDispatcher::AsyncDispatcher(LogSink log_sink, std::size_t worker_count, std::size_t capacity)
    : log_sink_(std::move(log_sink)), capacity_(capacity) {
  workers_.reserve(worker_count);
  for (std::size_t index = 0; index < worker_count; ++index) {
    workers_.emplace_back([this] { Work(); });
  }
}

AsyncDispatcher::~AsyncDispatcher() {
  Close(kCloseDrainTimeout);
  for (auto& worker : workers_) {
    if (worker.joinable()) {
      worker.join();
    }
  }
}

bool AsyncDispatcher::Submit(Task task) {
  std::optional<std::uint64_t> dropped;
  {
    const std::lock_guard<std::mutex> lock(mutex_);
    if (!accepting_) {
      return false;
    }
    if (tasks_.size() >= capacity_) {
      tasks_.pop_front();
      dropped = drop_logger_.RecordDrop();
    }
    tasks_.push_back(std::move(task));
  }
  changed_.notify_one();

  if (dropped.has_value() && log_sink_) {
    log_sink_(LogLevel::kWarning,
              "oppex: dropped " + std::to_string(*dropped) + " incidents in the last minute");
  }
  return true;
}

void AsyncDispatcher::Close(std::chrono::milliseconds timeout) noexcept {
  std::size_t abandoned = 0;
  {
    std::unique_lock<std::mutex> lock(mutex_);
    if (!accepting_) {
      return;
    }
    accepting_ = false;

    changed_.wait_for(lock, timeout, [this] { return tasks_.empty() && active_ == 0; });

    // Stops the workers whether or not the queue emptied. A thread cannot be
    // interrupted, so a worker blocked in an HTTP attempt finishes that attempt
    // on its own; the attempt timeout already bounds how long that takes.
    draining_ = false;
    abandoned = tasks_.size();
    tasks_.clear();
  }
  changed_.notify_all();

  // Reported once and directly, rather than through the rate-limited overflow
  // counter: a shutdown loss and an overload loss are different events, and a
  // rate-limited counter's last batch can go unreported.
  if (abandoned > 0 && log_sink_) {
    log_sink_(LogLevel::kWarning, "oppex: force-dropped " + std::to_string(abandoned) +
                                      " pending incidents during close");
  }
}

void AsyncDispatcher::Work() {
  while (true) {
    Task task;
    {
      std::unique_lock<std::mutex> lock(mutex_);
      changed_.wait(lock, [this] { return !tasks_.empty() || !draining_; });
      if (tasks_.empty()) {
        return;
      }
      task = std::move(tasks_.front());
      tasks_.pop_front();
      ++active_;
    }

    // A task owns its own error handling; see Client::Submit, which wraps every
    // task in a catch-all before it reaches here.
    task();

    {
      const std::lock_guard<std::mutex> lock(mutex_);
      --active_;
    }
    changed_.notify_all();
  }
}

}  // namespace oppex
